### PR TITLE
Store

### PR DESCRIPTION
**Pull request by tok-kkk**
Move the store into its own package and import in the lightnode.

### CODE REVIEW

File: p2p/p2p.go [hunk 1]
```diff
@@ -209,16 +208,29 @@ func (p2p *P2P) sendRequest(request jsonrpc.JSONRequest, multi peer.MultiAddr) j
 
 // handleQueryPeers retrieves at most 5 random multi-addresses from the store.
 func (p2p *P2P) handleQueryPeers(request jsonrpc.QueryPeersRequest) jsonrpc.Response {
-	addresses := p2p.randomPeers()
-	response := jsonrpc.QueryPeersResponse{
-		Peers: addresses,
+	var response jsonrpc.QueryPeersResponse
+	addresses, err := p2p.randomPeers()
+	if err != nil {
+		p2p.logger.Errorf("fail to get random peers from the multiAddress store %v", err)
```
Comment (jazg) [inline p2p/p2p.go]: "failed to get random peers from the multi-address store: %v"

File: p2p/p2p.go [hunk 2]
```diff
@@ -227,50 +239,36 @@ func (p2p *P2P) randomPeers() []string {
...
 }
 
 // handleQueryNumPeers retrieves the number of multi-addresses in the store.
 func (p2p *P2P) handleQueryNumPeers(request jsonrpc.QueryNumPeersRequest) jsonrpc.Response {
-	return jsonrpc.QueryNumPeersResponse{
-		NumPeers: p2p.store.MultiAddressEntries(),
+	var response jsonrpc.QueryNumPeersResponse
+	addresses, err := p2p.store.MultiAddrs()
+	if err != nil {
+		p2p.logger.Errorf("fail to get all peers from the multiAddress store %v", err)
```
Comment (jazg) [inline p2p/p2p.go]: "failed to get peers from the multi-address store: %v"

File: p2p/p2p.go [hunk 3]
```diff
@@ -227,50 +239,36 @@ func (p2p *P2P) randomPeers() []string {
...
-		response := jsonrpc.QueryStatsResponse{
-			Version: p2p.health.Version(),
-		}
-		cpus, err := p2p.health.CPUs()
+		var response jsonrpc.QueryStatsResponse
+		info, err := p2p.health.Info()
 		if err != nil {
-			response.Error = ErrFailedToGetStatistics
-			return response
+			p2p.logger.Errorf("fail to get self health info, %v", err)
```
Comment (jazg) [inline p2p/p2p.go]: "failed to get lightnode health information: %v"

File: rpc/client.go [hunk 4]
```diff
@@ -68,9 +71,9 @@ func (client *Client) invoke(message InvokeRPC) tau.Message {
 	case jsonrpc.SendMessageRequest:
 		return client.handleMessage(request, jsonrpc.MethodSendMessage, message.Addresses)
 	case jsonrpc.ReceiveMessageRequest:
-		// Check if the message already exists in the store and if so, write it to the responder channel.
-		response, err := client.store.Message(request.MessageID)
-		if err != nil {
+		// Check if the message already exists in the multiStore and if so, write it to the responder channel.
```
Comment (jazg) [inline rpc/client.go]: "messages store" rather than "multiStore"

File: rpc/client.go [hunk 5]
```diff
@@ -118,9 +121,9 @@ func (client *Client) runWorkers(n int) {
 					continue
 				}
 
-				// Write result to store.
+				// Write result to multiStore.
 				if err := client.insertMessageResult(call.Request, resp); err != nil {
-					client.logger.Errorf("cannot store the ReceiveMessageResponse: %v", err)
+					client.logger.Errorf("cannot multiStore the ReceiveMessageResponse: %v", err)
```
Comment (jazg) [inline rpc/client.go]: "cannot store the ReceiveMessageResponse in the messages store: %v"

File: rpc/client.go [hunk 6]
```diff
@@ -118,9 +121,9 @@ func (client *Client) runWorkers(n int) {
 					continue
 				}
 
-				// Write result to store.
+				// Write result to multiStore.
```
Comment (jazg) [inline rpc/client.go]: "messages store" rather than "multiStore"

File: rpc/client.go [hunk 7]
```diff
@@ -160,8 +163,8 @@ func (client *Client) handleMessage(request jsonrpc.Request, method string, addr
 
 // handleRequest sending the constructed request to the queue.
 func (client *Client) handleRequest(method string, data []byte, address addr.Addr, results chan jsonrpc.Response) error {
-	// Get multi-address of the darknode from store.
-	multi, err := client.store.MultiAddress(address)
+	// Get multi-address of the darknode from multiStore.
```
Comment (jazg) [inline rpc/client.go]: "Get multi-address of the Darknode from the multi-address store."

File: rpc/server.go [hunk 8]
```diff
@@ -79,7 +79,7 @@ func NewSendMessage(req jsonrpc.Request) SendMessage {
 }
 
 // QueryMessage is created and propagated by the server to its parent when receiving a QueryPeers or QueryNumPeers
-// request. These messages get forwarded to the P2P task which queries the store and writes the response in the
+// request. These messages get forwarded to the P2P task which queries the multiStore and writes the response in the
```
Comment (jazg) [inline rpc/server.go]: "queries the multi-address store"

File: rpc/client_test.go [hunk 9]
```diff
@@ -299,18 +289,16 @@ var _ = Describe("RPC client", func() {
 			badServer := initTimeoutServer()
 			defer badServer.Close()
 
-			// Construct the store for the client.
+			// Construct the multiStore for the client.
```
Comment (jazg) [inline rpc/client_test.go]: "Construct the multi-address store for the client."

File: testutils/darknode.go [hunk 10]
```diff
@@ -63,9 +63,7 @@ func (node *MockDarknode) Run(done <-chan struct{}) {
 		server.Close()
 	}()
 
-	if err := server.ListenAndServe(); err != nil {
-		log.Printf("failed to serve: %v", err)
-	}
+	server.ListenAndServe()
```
Comment (jazg) [inline testutils/darknode.go]: Reason why the error check was removed?

Comment (tok-kkk) [reply testutils/darknode.go]: The `ListenAndServe` will always return a non-nil err and I think it's ok to ignore the error.

File: rpc/client.go [hunk 11]
```diff
@@ -273,28 +267,45 @@ Loop:
...
-func (client Client) insertMessageResult(jsonRequest jsonrpc.JSONRequest, response jsonrpc.ReceiveMessageResponse) error {
-	var request jsonrpc.ReceiveMessageRequest
-	if err := json.Unmarshal(jsonRequest.Params, &request); err != nil {
-		return err
+func (client *Client) getMessage(id string) (jsonrpc.ReceiveMessageResponse, error) {
+	response, ok := client.messageStore[id]
+	if !ok {
+		return response, errors.New("response not available")
+	}
+	if time.Now().Sub(client.lastSeen[id]) > 5*time.Minute {
```
Comment (jazg) [inline rpc/client.go]: Can we make this timeout configurable?

**Pull request merged**